### ISSUE TITLE
Segfault when decoding possible bad data

### ISSUE THREAD

**Issue by JDarnley** [bbc/vc2hqdecode#16]
I encountered a segfault while attempting to decode decoding a stream of our own encoder.  Like in issue no. 15 I am not saying that the bitstream is correct but crashing is not good behaviour.

Arch Linux, x86-64, kernel 4.9.78-1, Boost 1.66.0-1, gcc 7.2.1+20180116-1

gdb backtrace when I hit Ctrl+C
```
GNU gdb (GDB) 8.0.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/james/src/vc2hqdecode/testprogs/vc2decode...done.
(gdb) r
Starting program: /home/james/src/vc2hqdecode/testprogs/vc2decode temp.vc2 /dev/null
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
INFO: Processor Features:
INFO:   SSE4.2 [ ]
INFO:   AVX    [ ]
INFO:   AVX2   [ ]
Read 836366bytes of input data
INFO: Processing Sequence Header
INFO: Configuring for 1920 x 1080
INFO: Processing Transform Params
INFO: Configuring for 1 threads, 4 jobs

Program received signal SIGSEGV, Segmentation fault.
0x00005555555a69f9 in LeGall_5_3_invtransform_H_final_1<10, short> (_idata=0x7ffff5861010, istride=512, odata=0x7ffff6c11c10 "", ostride=960, iwidth=496, iheight=552, ooffset_x=0,
    ooffset_y=<optimized out>, owidth=480, oheight=544) at legall_invtransform.hpp:268
268             ((uint16_t *)odata)[((y - ooffset_y)*ostride + x - ooffset_x + 0*skip)] = (uint16_t)MIN(MAX((((X   + 1) >> 1) + offset), 0), clip);
(gdb) bt
#0  0x00005555555a69f9 in LeGall_5_3_invtransform_H_final_1<10, short> (_idata=0x7ffff5861010, istride=512, odata=0x7ffff6c11c10 "", ostride=960, iwidth=496, iheight=552, ooffset_x=0,
    ooffset_y=<optimized out>, owidth=480, oheight=544) at legall_invtransform.hpp:268
#1  0x0000555555576c56 in VC2Decoder::Decode (_ostride=0x7fffffffe630, _odata=0x5555557d9ef0, job=0x55555580dc50, this=0x5555557da2a0) at VC2Decoder.cpp:1549
#2  VC2Decoder::decodeFrame (this=this@entry=0x5555557da2a0, _idata=_idata@entry=0x7ffff7f0e053 "", ilength=ilength@entry=836299, odata=odata@entry=0x5555557d9ef0,
    ostride=ostride@entry=0x7fffffffe630) at VC2Decoder.cpp:1055
#3  0x000055555557c0f2 in VC2Decoder::sequenceDecodeOnePicture (this=0x5555557da2a0, _idata=0x7fffffffdf18, ilength=<optimized out>, odata=0x5555557d9ef0, ostride=0x7fffffffe630,
    skip_aux=true) at VC2Decoder.cpp:268
#4  0x000055555556c904 in vc2decode_decode_one_picture (handle=<optimized out>, idata=<optimized out>, ilength=<optimized out>, odata=<optimized out>, ostride=<optimized out>,
    skip_aux=<optimized out>) at vc2hqdecode.cpp:117
#5  0x000055555555b332 in main (argc=<optimized out>, argv=<optimized out>) at vc2decode.cpp:309
```
Sample is available here https://0x0.st/sTz4.vc2

**Comment by jamesba** [comment]
Interestingly this doesn't segfault if the decoder has been built with CXXFLAGS="-fsanitize=address" LDFLAGS="-lasan"

**Comment by hannob** [comment]
@jamesba for me it does:

```
==27490==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f54ffcff800 at pc 0x55bfcd7a8d5a bp 0x7ffeee7d5b70 sp 0x7ffeee7d5b60
WRITE of size 16 at 0x7f54ffcff800 thread T0
    #0 0x55bfcd7a8d59 in _mm_storeu_si128(long long __vector(2)*, long long __vector(2)) /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/emmintrin.h:720
    #1 0x55bfcd7a8d59 in void LeGall_5_3_invtransform_H_final_1_10_sse4_2_int16_t<10>(void*, int, char const*, int, int, int, int, int, int, int) /f/vc2/vc2hqdecode/vc2inversetransform_sse4_2/legall_invtransform.hpp:364
    #2 0x55bfcd7002cf in VC2Decoder::Decode(JobData*, unsigned short**, int*) /f/vc2/vc2hqdecode/vc2hqdecode/VC2Decoder.cpp:1549
    #3 0x55bfcd7002cf in VC2Decoder::decodeFrame(char*, int, unsigned short**, int*) /f/vc2/vc2hqdecode/vc2hqdecode/VC2Decoder.cpp:1055
    #4 0x55bfcd710af3 in VC2Decoder::sequenceDecodeOnePicture(char**, int, unsigned short**, int*, bool) /f/vc2/vc2hqdecode/vc2hqdecode/VC2Decoder.cpp:268
    #5 0x55bfcd6c7d5b in vc2decode_decode_one_picture /f/vc2/vc2hqdecode/vc2hqdecode/vc2hqdecode.cpp:117
    #6 0x55bfcd66dd4c in main /f/vc2/vc2hqdecode/testprogs/vc2decode.cpp:311
    #7 0x7f5502be7f85 in __libc_start_main (/lib64/libc.so.6+0x20f85)
    #8 0x55bfcd674be9 in _start (/r/vc2/vc2decode+0x1abe9)

0x7f54ffcff800 is located 0 bytes to the right of 8294400-byte region [0x7f54ff516800,0x7f54ffcff800)
allocated by thread T0 here:
    #0 0x7f5503c05e20 in malloc (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xe0e20)
    #1 0x55bfcd66dfa3 in main /f/vc2/vc2hqdecode/testprogs/vc2decode.cpp:278
    #2 0x7f5502be7f85 in __libc_start_main (/lib64/libc.so.6+0x20f85)
```

**Comment by jamesba** [comment]
Fixed in commit 9560ab35be149526966004997bc079b99f47ec99
